properties/serialization.go

```diff
@@ -0,0 +1,286 @@
+	var data bytes.Buffer
+
+	for key, value := range m {
+		keySize := uint32(len(key))
+		binary.Write(&sizeData, binary.LittleEndian, keySize)
+		keyData := *(*[]byte)(unsafe.Pointer(&key))
+		data.Write(keyData)
+		data.WriteByte(0)
+		valueSize := uint32(len(value))
+		binary.Write(&sizeData, binary.LittleEndian, valueSize)
+		valueData := *(*[]byte)(unsafe.Pointer(&value))
+		data.Write(valueData)
+		data.WriteByte(0)
+	}
+
+	// Combine size data and actual data into the main buffer
+	buf.Write(sizeData.Bytes())
+	buf.Write(data.Bytes())
+
+	return buf.Bytes()
+}
+
+// deserializeStringMap deserializes the bytes to key value map, used for string only type maps
+//   - keys are always string
+//   - value are always string
+func deserializeStringMap(bs []byte) map[string]string {
+	numHeaders := binary.LittleEndian.Uint32(bs[0:4])
+	var sizeIndex = 4
+	var dataIndex = 4 + 4*2*int(numHeaders)
+	ret := make(map[string]string, numHeaders)
+	for i := 0; i < int(numHeaders); i++ {
+		keySize := int(binary.LittleEndian.Uint32(bs[sizeIndex : sizeIndex+4]))
+		sizeIndex += 4
+		keyPtr := bs[dataIndex : dataIndex+keySize]
+		key := *(*string)(unsafe.Pointer(&keyPtr))
+		dataIndex += keySize + 1
+
+		valueSize := int(binary.LittleEndian.Uint32(bs[sizeIndex : sizeIndex+4]))
+		sizeIndex += 4
+		valuePtr := bs[dataIndex : dataIndex+valueSize]
+		value := *(*string)(unsafe.Pointer(&valuePtr))
+		dataIndex += valueSize + 1
+		ret[key] = value
+	}
+	return ret
+}
+
+// serializeStringSlice serializes a slice of strings into a single byte slice.
+// The resulting byte slice can be used for efficient storage or transmission.
+// Each string in the input is prefixed with its length, allowing for efficient deserialization.
+func serializeStringSlice(strings []string) []byte {
+	totalSize := 4
+	for _, str := range strings {
+		totalSize += 8 + len(str) + 2
+	}
+	bs := make([]byte, totalSize)
+	binary.LittleEndian.PutUint32(bs[:4], uint32(len(strings)))
+	idx := 4
+	dataIdx := 4 + 8*len(strings)
+	for _, str := range strings {
+		binary.LittleEndian.PutUint64(bs[idx:idx+8], uint64(len(str)))
+		idx += 8
+		copy(bs[dataIdx:dataIdx+len(str)], str)
+		dataIdx += len(str) + 2
+	}
+	return bs
+}
+
+// deserializeStringSlice deserializes the given byte slice to string slice.
+func deserializeStringSlice(bs []byte) []string {
+	numStrings := int(binary.LittleEndian.Uint32(bs[:4]))
+	ret := make([]string, numStrings)
+	idx := 4
+	dataIdx := 4 + 8*numStrings
+	for i := 0; i < numStrings; i++ {
+		strLen := int(binary.LittleEndian.Uint64(bs[idx : idx+8]))
+		idx += 8
+		ret[i] = string(bs[dataIdx : dataIdx+strLen])
+		dataIdx += strLen + 2
+	}
+	return ret
+}
+
+// serializeTimestamp serializes the given timestamp to bytes.
+// The resulting byte slice can be used for efficient storage or transmission.
+func serializeTimestamp(timestamp time.Time) []byte {
+	nanos := timestamp.UnixNano()
+	bs := make([]byte, 8)
+	binary.LittleEndian.PutUint64(bs, uint64(nanos))
+	return bs
```

Comment:
I believe this can be the array instead of slice (always better to avoid allocation)

```suggestion
	nanos := timestamp.UnixNano()
	var bs [8]byte
	binary.LittleEndian.PutUint64(bs[:], uint64(nanos))
	return bs
```

Reply:
Did not fix this as it would require changes in the test framework: `WithProperty` does only allow for []byte, not [8]byte. I do not see much value add, to force this to be the only bounded array of size 8, across an API with []byte slices only.

```golang
func TestGetPropertyTimestamp(t *testing.T) {
	now := time.Now().UTC()

	opt := proxytest.NewEmulatorOption().WithProperty([]string{"someTimestampPath"}, serializeTimestamp(now))
	_, reset := proxytest.NewHostEmulator(opt)
	defer reset()	

	result, err := getPropertyTimestamp([]string{"someTimestampPath"})
	require.NoError(t, err)
	require.Equal(t, now, result)
}
```